Add epidatr blog post (#912)
cmu-delphi/www-main: content/blog/2023-09-27-epidatr.Rmd

---
title: New package released! Epidatr, an R client for Delphi Epidata API
author: Dmitry Shemetov, David Weber
date: 2023-09-27
tags:
  - r
  - epidata
  - covidcast
authors:
  - dmitry
  - davidweb
heroImage: blog-lg-epidatr.jpg
heroImageThumb: blog-thumb-epidatr.jpg
summary: |
  `epidatr` is designed to streamline the downloading and usage of data from the [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/). It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format.

output:
  blogdown::html_page:
    toc: true
---

> The [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/) provides real-time access to epidemiological surveillance data for influenza, COVID-19, and other diseases from both official government sources such as the [Center for Disease Control (CDC)](https://www.cdc.gov/datastatistics/index.html), private partners such as [Facebook](https://delphi.cmu.edu/blog/2020/08/26/covid-19-symptom-surveys-through-facebook/) and [Change Healthcare](https://www.changehealthcare.com/), and other public datasets like [Google Trends](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets/covid19-search-trends). It is built and maintained by the Carnegie Mellon University [Delphi research group](https://delphi.cmu.edu/).

Reviewer: ```suggestion
The [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/) provides real-time access to epidemiological surveillance data for influenza, COVID-19, and other diseases from both official government sources such as the [Centers for Disease Control and Prevention (CDC)](https://www.cdc.gov/datastatistics/index.html), private partners such as [Meta](https://delphi.cmu.edu/blog/2020/08/26/covid-19-symptom-surveys-through-facebook/) and [Change Healthcare](https://www.changehealthcare.com/), and other public datasets like [Google Trends](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets/covid19-search-trends). It is built and maintained by the Carnegie Mellon University [Delphi Research Group](https://delphi.cmu.edu/).
```

Additionally, Change Healthcare has merged into Optum, but "Change Healthcare" might still count as a "current" name, based on [https://www.changehealthcare.com/](https://www.changehealthcare.com/).  I didn't include any change in the suggestion above, since it's probably easier for readers to search the API docs for Change and get the right endpoint.

Reviewer 2: These make sense, though i might say something like "Facebook (now Meta)" since they were Facebook when we partnered with them and because we refer to them as such in the linked article.

Today we introduce the R package `epidatr`, available [on CRAN](https://cloud.r-project.org/web/packages/epidatr/index.html), with the source and development [on github](https://github.com/cmu-delphi/epidatr).

This package is designed to streamline the downloading and usage of data from the Delphi Epidata API. It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format. The API stores a historical record of all data, including corrections and updates, which is particularly useful for accurately backtesting forecasting models. We also provide packages for downstream data processing ([epiprocess](https://github.com/cmu-delphi/epiprocess)) and modeling ([epipredict](https://github.com/cmu-delphi/epipredict)).

## Usage

```
library(epidatr)

> # Obtain the smoothed covid-like illness (CLI) signal from the Facebook survey as it was on April 10, 2021 for the US

Reviewer: ```suggestion
# Obtain the smoothed covid-like illness (CLI) signal from the Delphi's US
# COVID-19 Trends and Impact Survey (CTIS), in partnership with Facebook, as it
# was reported on April 10, 2021, for the US at a national level
```

trying to get closer to the language [here](https://www.pnas.org/doi/full/10.1073/pnas.2111454118).  Not sure about suggesting "Meta" above & "Facebook" here though

 2 us        smoothed_cli fb-su… nation   day       2021-01-02
 3 us        smoothed_cli fb-su… nation   day       2021-01-03
 4 us        smoothed_cli fb-su… nation   day       2021-01-04
 5 us        smoothed_cli fb-su… nation   day       2021-01-05
 6 us        smoothed_cli fb-su… nation   day       2021-01-06
 7 us        smoothed_cli fb-su… nation   day       2021-01-07
 8 us        smoothed_cli fb-su… nation   day       2021-01-08
 9 us        smoothed_cli fb-su… nation   day       2021-01-09
10 us        smoothed_cli fb-su… nation   day       2021-01-10
# ℹ 141 more rows
# ℹ 9 more variables: direction <dbl>, issue <date>,
#   lag <dbl>, missing_value <dbl>, missing_stderr <dbl>,
#   missing_sample_size <dbl>, value <dbl>, stderr <dbl>,
#   sample_size <dbl>
# ℹ Use `print(n = ...)` to see more rows
```

## Installation

You can install the stable version of this package from CRAN:

```
install.packages("epidatr")
pak::pkg_install("epidatr")
renv::install("epidatr")
```

Or if you want the development version, install from GitHub:

```
# Install the dev version using `pak` or `remotes`
pak::pkg_install("cmu-delphi/epidatr")
remotes::install_github("cmu-delphi/epidatr")

> renv::install("cmu-delphi/epidatr")

Reviewer: Are we switching the default epidatr branch to main?  Or keeping as dev because it's already on CRAN?  If the former, then we'll want an `@dev` here

Reviewer 3: Probably safe to put @dev here, just in case we do switch. Will do.

```

### API Keys

The Delphi API requires a (free) API key for full functionality. To generate
your key, register for a pseudo-anonymous account
[here](https://api.delphi.cmu.edu/epidata/admin/registration_form) and see more
discussion on the [general API
website](https://cmu-delphi.github.io/delphi-epidata/api/api_keys.html). The
`epidatr` client will automatically look for this key in the R option
`delphi.epidata.key` or in the environment variable

> `DELPHI_EPIDATA_KEY`. We recommend storing your key in `.Renviron` file, which R

Reviewer: ```suggestion
`DELPHI_EPIDATA_KEY`. We recommend storing your key in a `.Renviron` file, which R
```

(or go further and recommend `~/.Renviron`?)

Reviewer 2: Good catch, this definitely needs an article!  I'm not sure if "a" or "an" is appropriate here -- i would use the former if pronouncing the "dot" and the latter if leaving it out...  Maybe we should use "the" or "your"?

will read by default.

Note that for the time being, the private endpoints (i.e. those prefixed with
`pvt`) will require a separate key that needs to be passed as an argument.

## For users of the covidcast R package

> The `epidatr` package is a complete rewrite of the [`covidcast` package](https://cmu-delphi.github.io/covidcast/covidcastR/), with a focus on speed, reliability, and ease of use. The `covidcast` package is deprecated and will no longer be updated.

Reviewer: should mention it has more endpoints / data sources than the covidcast package, and maybe direct to `pub_covidcast` if migrating

---

> title: New package released! `epidatr`, an R client for Delphi Epidata API

Reviewer 2: The string "\`epidatr\`" in the title renders literally, which is probably not what was intended.  I'm not sure what to do here.

Reviewer 3: Happy to just get rid of the ticks.

author: Dmitry Shemetov, David Weber
date: 2023-09-27
tags:
  - r
  - epidata
  - covidcast
authors:
  - dmitry

>   - david

Reviewer 2: This plain "david" is pointing to David Farrow in the preview (at the bottom of the page).  We need to add a key for David Weber in https://github.com/cmu-delphi/www-main/blob/dev/content/people/index.md, and then use it here.  We could also add biographical blurbs for them in that same file, if desired.
```suggestion
  - davidweb
```

heroImage: blog-lg-epidatr.jpg
heroImageThumb: blog-thumb-epidatr.jpg
summary: |
  `epidatr` is designed to streamline the downloading and usage of data from the [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/). It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format.
 
output:
  blogdown::html_page:
    toc: true
---

> The [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/) provides real-time access to epidemiological surveillance data for influenza, COVID-19, and other diseases from both official government sources such as the [Center for Disease Control (CDC)](https://www.cdc.gov/datastatistics/index.html) and [Google Trends](https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/google-symptoms.html) and private partners such as [Facebook](https://delphi.cmu.edu/blog/2020/08/26/covid-19-symptom-surveys-through-facebook/) and [Change Healthcare](https://www.changehealthcare.com/). It is built and maintained by the Carnegie Mellon University [Delphi research group](https://delphi.cmu.edu/).

Reviewer 2: The existing wording makes it sound like Google Trends is an official government source...  We should probably also use a more official link for Google Trends, to make it more like the links around it:

```suggestion
The [Delphi Epidata API](https://cmu-delphi.github.io/delphi-epidata/) provides real-time access to epidemiological surveillance data for influenza, COVID-19, and other diseases from both official government sources such as the [Center for Disease Control (CDC)](https://www.cdc.gov/datastatistics/index.html), private partners such as [Facebook](https://delphi.cmu.edu/blog/2020/08/26/covid-19-symptom-surveys-through-facebook/) and [Change Healthcare](https://www.changehealthcare.com/), and other public datasets like [Google Trends](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets/covid19-search-trends). It is built and maintained by the Carnegie Mellon University [Delphi research group](https://delphi.cmu.edu/).
```

> 

Reviewer 2: We need some actual links to epidatr!!!
```suggestion

Today we introduce the R package "`epidatr`", available [on CRAN](https://cloud.r-project.org/web/packages/epidatr/index.html), with the source and development [on github](https://github.com/cmu-delphi/epidatr).

```

Author: Links to the tool we're talking about? That's crazy!

> This package is designed to streamline the downloading and usage of data from the [Delphi Epidata
> API](https://cmu-delphi.github.io/delphi-epidata/). It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format. The API stores a historical record of all data, including corrections and updates, which is particularly useful for accurately backtesting forecasting models. We also provide packages for downstream data processing ([epiprocess](https://github.com/cmu-delphi/epiprocess)) and modeling ([epipredict](https://github.com/cmu-delphi/epipredict)).

Reviewer 2: We already provided the same link above (also remove the superfluous newline)
```suggestion
This package is designed to streamline the downloading and usage of data from the Delphi Epidata API. It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format. The API stores a historical record of all data, including corrections and updates, which is particularly useful for accurately backtesting forecasting models. We also provide packages for downstream data processing ([epiprocess](https://github.com/cmu-delphi/epiprocess)) and modeling ([epipredict](https://github.com/cmu-delphi/epipredict)).
```

  as_of = "2021-06-01")
epidata
```

```
# A tibble: 6 × 15
  geo_value signal       source   geo_type time_type time_value
  <chr>     <chr>        <chr>    <fct>    <fct>     <date>
1 us        smoothed_cli fb-surv… nation   day       2021-04-05
2 us        smoothed_cli fb-surv… nation   day       2021-04-06
3 us        smoothed_cli fb-surv… nation   day       2021-04-07
4 us        smoothed_cli fb-surv… nation   day       2021-04-08
5 us        smoothed_cli fb-surv… nation   day       2021-04-09
6 us        smoothed_cli fb-surv… nation   day       2021-04-10
# ℹ 9 more variables: direction <dbl>, issue <date>,
#   lag <int>, missing_value <int>, missing_stderr <int>,
#   missing_sample_size <int>, value <dbl>, stderr <dbl>,
#   sample_size <dbl>
```

## Installation

You can install the stable version of this package from CRAN:

``` r
install.packages("epidatr")
pak::pkg_install("epidatr")
renv::install("epidatr")
```

Or if you want the development version, install from GitHub:

``` r
# Install the dev version using `pak` or `remotes`
pak::pkg_install("cmu-delphi/epidatr")
remotes::install_github("cmu-delphi/epidatr")
renv::install("cmu-delphi/epidatr")

> 

Reviewer 2: The code block isnt properly terminated
```suggestion
\```

```
(remove the backslash; its just there to get get the suggestion to render properly)

Today we introduce the R package "`epidatr`", available [on CRAN](https://cloud.r-project.org/web/packages/epidatr/index.html), with the source and development [on github](https://github.com/cmu-delphi/epidatr).

This package is designed to streamline the downloading and usage of data from the Delphi Epidata API. It provides a simple R interface to the API, including functions for downloading data, parsing the results, and converting the data into a tidy format. The API stores a historical record of all data, including corrections and updates, which is particularly useful for accurately backtesting forecasting models. We also provide packages for downstream data processing ([epiprocess](https://github.com/cmu-delphi/epiprocess)) and modeling ([epipredict](https://github.com/cmu-delphi/epipredict)).

## Usage

> ```R

Reviewer 2: this code block seems to be automatically collapsed/folded/hidden, which is kinda silly because the article is for a package that people will use in their own code (and the resulting output block below it is not collapsed).

I think we can fix this by removing the `R` at the top of this block, or by replacing the `R` with `{r class.source = 'fold-show'}` as described [here](https://bookdown.org/yihui/rmarkdown-cookbook/fold-show.html).  The former is much simpler and i think i would go with that.

Reviewer 3: Tried this with no luck. I have a feeling that we're not really using blogdown, but instead just rendering html that is then passed through our Hugo blog's styling.

library(epidatr)
# Obtain the smoothed covid-like illness (CLI) signal from the Facebook survey as it was on April 10, 2021 for the US
epidata <- covidcast(
  source = "fb-survey",
  signals = "smoothed_cli",
  geo_type = "nation",
  time_type = "day",
  geo_values = "us",
  time_values = epirange(20210101, 20210601),
  as_of = "2021-06-01")
epidata
```

```
# A tibble: 6 × 15
  geo_value signal       source   geo_type time_type time_value
  <chr>     <chr>        <chr>    <fct>    <fct>     <date>
1 us        smoothed_cli fb-surv… nation   day       2021-04-05
2 us        smoothed_cli fb-surv… nation   day       2021-04-06
3 us        smoothed_cli fb-surv… nation   day       2021-04-07
4 us        smoothed_cli fb-surv… nation   day       2021-04-08
5 us        smoothed_cli fb-surv… nation   day       2021-04-09
6 us        smoothed_cli fb-surv… nation   day       2021-04-10
# ℹ 9 more variables: direction <dbl>, issue <date>,
#   lag <int>, missing_value <int>, missing_stderr <int>,
#   missing_sample_size <int>, value <dbl>, stderr <dbl>,
#   sample_size <dbl>
```

## Installation

You can install the stable version of this package from CRAN:

> ``` r

Reviewer 2: Same problem as described above, i think this should be un-collapsed by default.

install.packages("epidatr")
pak::pkg_install("epidatr")
renv::install("epidatr")
```

Or if you want the development version, install from GitHub:

> ``` r

Reviewer 2: Same problem as described above, i think this should be un-collapsed by default.

# Install the dev version using `pak` or `remotes`
pak::pkg_install("cmu-delphi/epidatr")
remotes::install_github("cmu-delphi/epidatr")
renv::install("cmu-delphi/epidatr")

> \```

Reviewer 2: we gotta remove this backslash.  ill do that and re-render the html file when i have a moment

Author: THANK YOU!

Reviewer 2: Remember how i said above "I added the rendered html file... It was a pain to get it to work, but i should be able to easily do it again in the future!"?  Well, i forgot how and didnt write it down, so i had to figure out how to do it again 😡 ...  But for posterity, i downloaded and extracted a pandoc binary (from [here](https://github.com/jgm/pandoc/releases/tag/3.1.11.1)), and added its `bin/` directory to my `PATH` before executing `Rscript -e 'blogdown::build_site(...`

### API Keys

The Delphi API requires a (free) API key for full functionality. To generate
your key, register for a pseudo-anonymous account
[here](https://api.delphi.cmu.edu/epidata/admin/registration_form) and see more
discussion on the [general API
website](https://cmu-delphi.github.io/delphi-epidata/api/api_keys.html). The
`epidatr` client will automatically look for this key in the R option
`delphi.epidata.key` or in the environment variable
`DELPHI_EPIDATA_KEY`. We recommend storing your key in `.Renviron` file, which R
will read by default.

Note that for the time being, the private endpoints (i.e. those prefixed with
`pvt`) will require a separate key that needs to be passed as an argument.

> ## For users of the `covidcast` R package

Reviewer 2: This formatting screws with the layout of the "outline" element (the pinkish sorta table of contents on the left side of the blog post)...  I think we can just remove the fixed-width formatting, but putting "covidcast" in quotes would work too.
```suggestion
## For users of the covidcast R package
```
